### ISSUE TITLE
Improve SecretConnection/STS KDF?

### ISSUE THREAD

**Issue by tarcieri** [tendermint/tendermint#1165]
SecretConnection presently uses the NaCl `crypto_secretbox` construction for authenticated encryption:

https://github.com/tendermint/tendermint/blob/master/p2p/secret_connection.go#L131

`crypto_secretbox` decomposes to the following:

```
Salsa20Poly1305(HSalsa20(k, nonce), Ø{8}, frame)
```

The `crypto_secretbox` API imposes a 24-byte length requirement on the nonce, and for that reason SecretConnection first pre-hashes all of the cryptographic parameters with RIPEMD160:

https://github.com/tendermint/tendermint/blob/master/p2p/secret_connection.go#L131

So the overall construction ends up looking like this:

```
Salsa20Poly1305(HSalsa20(k, RIPEMD160(lowPubKey || hiPubKey)), Ø{8}, frame)
```

(glossing over the sender/receiver nonce tweaks momentarily)

So, funny thing, my understanding is HSalsa20 is just a PRF and can accept variable length input. The 24-byte restriction is a totally arbitrary one imposed by the `crypto_secretbox` API. So if you wanted to get rid of RIPEMD160 ([see discussion here](https://github.com/cosmos/cosmos-sdk/issues/263#issuecomment-359580833)), you could just do this:

```
Salsa20Poly1305(HSalsa20(k, lowPubKey || hiPubKey), Ø{8}, frame)
```

Tada, no need for RIPEMD160. If you step outside the constraints of the `crypto_secretbox` API, you should just be able to pass everything directly to `HSalsa20`. This should provide a 2<sup>128</sup> security level, as opposed to the 2<sup>80</sup> one provided by RIPEMD160.

All that said, I would probably suggest using HKDF instead of HSalsa20 here. See some of the criticism CodesInChaos makes about the similar nonce-based approach used by CurveCP:

https://codesinchaos.wordpress.com/2012/09/09/curvecp-1/

A more traditional way to build an authenticated key exchange is to compute a transcript hash of the entire key exchange and use that as an input to something like HKDF, and using that to derive symmetric keys.

HKDF is an extremely boring construction with well-understood, provable security properties created by Hugo Krawczyk, who just so happened to be this year's Levchin Award winner at RealWorldCrypto. HKDF is used by both TLS and Noise for key derivation.

With HKDF, instead of tweaking the nonce and using that to derive two keys, you can just ask for two symmetric keys worth of output, and have the conversation participants pick the appropriate key for their side of the conversation.

HKDF for Go is available here:

https://godoc.org/golang.org/x/crypto/hkdf

A minimalistic change to HKDF might change the above to something like this:

```
derivedKey = HKDFSHA256(k, lowPubKey || hiPubKey, "COSMOS_SECRETCONNECTION_KDF")
Salsa20Poly1305(derivedKey, Ø{8}, frame)
```

**Comment by zmanian** [comment]
I'm interested in taking this.

**Comment by ebuchman** [comment]
Linking to https://github.com/tendermint/tendermint/issues/1124 which suggests moving from xsalsa20poly1305 to xchacha20poly1305 - as implemented in https://github.com/tendermint/go-crypto/pull/98

The question here is whether to switch that to a custom hkdf-chacha20poly1305 -  is it really worth it ? the difference is basically that we get 2^256 msgs per key rather than 2^192 ?

**Comment by zmanian** [comment]
The alternative would be to start the nonce counters from zero and terminate the connection when they hit 2^192 -1 ?

**Comment by ebuchman** [comment]
We'll never hit that though right ? We couldn't send that many msgs with one key if we tried ...

And the nonce incrementer just wraps around. I feel like I'm missing something

**Comment by tarcieri** [comment]
@ebuchman the advantages of HKDF are:

* **Extended "salt" space**: by "fiat" XChaCha20 takes a 24-byte nonce (when really the nonce is just a parameter to the variable-length HChaCha20 PRF), which was the original impetus for using RIPEMD160 (as noted earlier). HKDF has no limitations on the length of the cryptographic parameters you can bind, so you can use it to bind whatever parameters you can express in some form of canonical string
* **Extra "info" parameter**: this parameter is mixed into every iteration of every derivation and is great for a "personalization string", i.e. some short string that "colors" a particular derivation for a particular purpose, such as deriving a bunch of purpose-specific keys from a single master key
* **Security proofs**: HKDF is used almost universally in this particular space because it has well-understood security proofs and therefore is exceedingly boring for this purpose. See its use in both TLS and Noise

**Comment by ebuchman** [comment]
OK - thanks for the extra details!

**Comment by ebuchman** [comment]
So - we just merged the move to xchacha20poly1305 by copying hchacha20 from github.com/aead/chacha20 and wrapping the stdlibs chacha20poly1305

But to implement this, we would remove hchacha20 and the sender/recvr nonces, and instead replace it with an HKDF stream from which we can just keep reading 40-byte chunks to pass as `nonce` into `Seal`, where the first 32-bytes are the key and the final 8 are the nonce for cc20p1305 ? 

And then to differentiate sender and receiver nonce input we would have two hkdf streams, something like:

```
derivedKeyRecv = HKDFSHA256(k, myPubKey || theirPubKey, "COSMOS_SECRETCONNECTION_KDF")
derivedKeySend = HKDFSHA256(k, theirPubKey || myPubKey, "COSMOS_SECRETCONNECTION_KDF")
```

Is that right ?

**Comment by ebuchman** [comment]
Uh, I'm not sure the HKDF output is unbounded, so maybe I'm still not understanding something

**Comment by tarcieri** [comment]
@ebuchman I'd suggest using HKDF to derive a unique key, rather than deriving a nonce. This is what the XSalsa and XChaCha constructions do (see above: the underlying Salsa/ChaCha nonce is `Ø{8}`).

The "X" versions of these constructions use HChaCha20 as a KDF to derive a unique key per nonce. You'd still use a nonce of `Ø{8}`, except the key would come from HKDF's output key material, rather than HChaCha20's output.

HKDF uses a method called expand-and-extract which is adaptable to the amount of output you want to generate. You can also change any of the parameters you pass to it (most notably the salt) to get a unique key per invocation.

**Comment by ebuchman** [comment]
Can you clarify what `Ø{8}` means ? 

The AEAD interface has a `Seal(dst, nonce, plaintext, additionalData []byte) []byte` method, and I thought we'd try to stick to that.

What I was trying to say was that what is called `nonce` in that method would here be the output from an HKDF that can be used as the key to ChaCha. And since that value would be unique with every message, the nonce used for the ChaCha can just be the empty bytes. Yes ? 

That said, we'd either need to seed an HKDF to give us large enough output or tweak the salt like we do the current nonces.

Also, it seems the X versions that take the 24-byte nonce use the first 16 bytes as a nonce in the HChaCha and the remaining 8 bytes as the nonce for the ChaCha, but I'm not sure if this is what you mean by `Ø{8}`.

Btw, thanks for taking the time to explain this stuff!

**Comment by tarcieri** [comment]
> Can you clarify what Ø{8} means ?

That's 8 * `\0` bytes passed as the 8-byte ChaCha20 nonce, after deriving a unique key using HChaCha20. But see below, I guess I was mistaken about this.

> since that value would be unique with every message, the nonce used for the ChaCha can just be the empty bytes. Yes ?

Correct

> That said, we'd either need to seed an HKDF to give us large enough output or tweak the salt like we do the current nonces.

You can invoke HKDF again with a different salt and derive a new key. So long as that isn't a performance concern (HKDF is pretty fast) it's a pretty nice option from a security perspective, IMO.

> Also, it seems the X versions that take the 24-byte nonce use the first 16 bytes as a nonce in the HChaCha and the remaining 8 bytes as the nonce for the ChaCha, but I'm not sure if this is what you mean by Ø{8}.

Yes, this is correct and I was wrong earlier: it uses the lower 8-bytes of the XChaCha nonce as the lower 8 bytes of the ChaCha20 nonce (with the upper 4-bytes set to `Ø{4}`)

**Comment by silasdavis** [comment]
+1 for this change, I have some other places I'd like to use this implementation. Could I also request that an exported (fine if it is internal/non-public to change) API is exposed?

**Comment by tarcieri** [comment]
Looking at #1563, and @ValarDragon's [branch here](https://github.com/tendermint/tendermint/tree/dev/update_compute_shared_secret), I think we're at (in pseudocode):

```
localPubKey, secKey = X25519KeyGen()
remotePubKey = [...]
lowPubKey, highPubKey = sort(localPubKey, remotePubKey)

nonce1 = HKDFSHA256(lowPubKey || highPubKey, Ø{32}, "TENDERMINT_SECRET_CONNECTION_NONCE_GEN", 24)
nonce2 = nonce1 ^ 0x01

if localPubKey == lowPubKey {
    recvNonce = nonce1
    sendNonce = nonce2
} else {
    recvNonce = nonce2
    sendNonce = nonce1
}

inputKeyMaterial = X25519(secKey, remotePubKey)
symmetricKey = HKDFSHA256(inputKeyMaterial, Ø{32}, "TENDERMINT_SECRET_CONNECTION_SHARED_SECRET_GEN", 32)

derivedKey[32] || derivedNonce[24] = HKDFSHA256(symmetricKey, sendNonce, "", 56)
ciphertext = ChaCha20Poly1305Encrypt(derivedKey, derivedNonce, msg)
```

That's 3 invocations of HKDF, which seems like 2 too many to me :wink: I would suggest eliminating the redundant invocations, since HKDF supports arbitrary length inputs for all parameters:

```
localPubKey, secKey = X25519KeyGen()
remotePubKey = [...]
lowPubKey, highPubKey = sort(localPubKey, remotePubKey)

nonce1 = lowPubKey || highPubKey
nonce2 = nonce1 ^ 0x01

if localPubKey == lowPubKey {
    recvNonce = nonce1
    sendNonce = nonce2
} else {
    recvNonce = nonce2
    sendNonce = nonce1
}

inputKeyMaterial = X25519(secKey, remotePubKey)

derivedKey[32] || derivedNonce[24] = HKDFSHA256(inputKeyMaterial, sendNonce, "TENDERMINT_SECRET_CONNECTION_SHARED_SECRET_GEN", 56)
ciphertext = ChaCha20Poly1305Encrypt(derivedKey, derivedNonce, msg)
```

I think you can do even better than this though, by simply deriving a unique key for sender and recipient:

```
localPubKey, secKey = X25519KeyGen()
remotePubKey = [...]
lowPubKey, highPubKey = sort(localPubKey, remotePubKey)
inputKeyMaterial = X25519(secKey, remotePubKey)

sendKey[32] || recvKey[32] = HKDFSHA256(inputKeyMaterial, lowPubKey || highPubKey, "TENDERMINT_SECRET_CONNECTION_SHARED_SECRET_GEN", 56)
ciphertext = ChaCha20Poly1305Encrypt(sendKey, Ø{12}, msg)
```

This would eliminate the existing `hash32` and `hash24` functions, the nonce tweaking, and even the need for any sort of nonce by deriving unique sender and recipient keys each time, and only perform one HKDF invocation.

**Comment by zmanian** [comment]
Right now the first 2 HKDF invocations are done once per connection and the 3rd HKDF invocation is done once per frame.

**Comment by tarcieri** [comment]
@zmanian aha, I see the `incr(2)Nonce` calls now:

https://github.com/tendermint/tendermint/blob/5b07b2c818630ad2527422aa71dfc0275d72b6a1/p2p/conn/secret_connection.go#L354

I'm a little confused exactly what that code is doing. It sort of looks like a counter, but I'm not sure.

I'd recommend a counter (starting from 0) as an implicit nonce passed directly to ChaCha20, which is pretty straightforward and what TLS, Noise, and CurveCP all use (or a combination of a pseudorandom nonce derived from HKDF and a counter). If you have that, you don't need to invoke HKDF for each frame (apologies for recommending the latter earlier, I didn't fully understand what was happening). Here's how Noise does it via the CipherState object:

http://noiseprotocol.org/noise.html#the-cipherstate-object

That should cut the number of HKDF invocations down to a single initial one to derive the sendKey and recvKey.

**Comment by ValarDragon** [comment]
Sorry for the super late response.

> I'm a little confused exactly what that code is doing. It sort of looks like a counter, but I'm not sure.
>
> I'd recommend a counter (starting from 0) as an implicit nonce passed directly to ChaCha20, which is pretty straightforward and what TLS, Noise, and CurveCP all use (or a combination of a pseudorandom nonce derived from HKDF and a counter). If you have that, you don't need to invoke HKDF for each frame (apologies for recommending the latter earlier, I didn't fully understand what was happening). Here's how Noise does it via the CipherState object:

We are just incrementing the nonce like a counter in the line you linked, and like what is done in noise from what I can gather. However to derive the subkey and chachaNonce, we're mimicking what xchacha20poly1305 did, just replacing hchacha with hkdf. (And removing the bit twiddling stuff) I assumed we need a hash function inside of seal, since we want this to be a drop-in replacement for xchacha20poly1305? While it doesn't matter for this use case, we'd need to support the ability to use randomized nonces for this to be a drop-in replacement for xchacha. (In case someone wants to use this in a stateless manner)

Disclaimer, I have not read much into any of the chacha schemes.

**Comment by tarcieri** [comment]
> We are just incrementing the nonce like a counter in the line you linked, and like what is done in noise from what I can gather.

I'm having trouble telling what that code is actually doing... why not just keep a `u64` counter which you `checked_add` to and then serializing it as little endian or big endian?

> While it doesn't matter for this use case, we'd need to support the ability to use randomized nonces for this to be a drop-in replacement for xchacha. (In case someone wants to use this in a stateless manner)

WireGuard uses XChaCha20 to support explicit nonces. But that's a pretty big change over a stream-oriented protocol capable of using implicit nonces: you'll need a sliding window and a slightly different wire format (one which carries an explicit nonce).

I'd recommend using ChaCha20 for implicit nonces, and crossing the XChaCha20/datagram-oriented bridge when you get there. In any case you can abstract across either usage pattern by having a stateful (i.e. `&mut`) type that picks the ChaCha20 nonce... it can be a simple counter for a stream oriented protocol, and CSPRNG + HChaCha20 for a stateful one.

**Comment by ValarDragon** [comment]
> I'm having trouble telling what that code is actually doing... why not just keep a u64 counter which you checked_add to and then serializing it as little endian or big endian?

Sounds like a good idea. I'll try to change this to be more readable during the week, as you suggested.

> Nobody uses XChaCha20 with NoiseSockets... it feels like a bit of a premature abstraction.

You're right. Just to make sure I understand correctly, the current state is to make secret connection just use chacha20poly1305, with a counter for the nonce. Since the nonce size is 96 bits, theres no concern of overflowing the nonce.

**Comment by ValarDragon** [comment]
Closing in favor of #2039
